algebird-core/src/main/scala/com/twitter/algebird/Batched.scala

```diff
@@ -0,0 +1,303 @@
+    that.foldLeft(this)((b, t) => b.combine(Batched(t)))
+
+  /**
+   * Provide an iterator over the underlying tree structure.
+   *
+   * This is the order used by `.sum`.
+   *
+   * This iterator traverses the tree from left-to-right. If the
+   * original expression was (w + x + y + z), this iterator returns w,
+   * x, y, and then z.
+   */
+  def iterator: Iterator[T] =
+    this match {
+      case Batched.Item(t) => Iterator.single(t)
+      case b => new Batched.ForwardItemsIterator(b)
+    }
+
+  /**
+   * Convert the batch to a `List[T]`.
+   */
+  def toList: List[T] =
+    reverseIterator.foldLeft(List.empty[T])((ts, t) => t :: ts)
+
+  /**
+   * Provide a reversed iterator over the underlying tree structure.
+   *
+   * This iterator traverses the tree from right-to-left. If the
+   * original expression was (w + x + y + z), this iterator returns z,
+   * y, x, and then w.
+   */
+  def reverseIterator: Iterator[T] =
+    this match {
+      case Batched.Item(t) => Iterator.single(t)
+      case b => new Batched.ReverseItemsIterator(b)
+    }
+
+  /**
+   * Report the size of the underlying tree structure.
+   */
+  def size: Int
+}
+
+object Batched {
+
+  /**
+   * Constructed a batch from a single value.
+   */
+  def apply[T](t: T): Batched[T] =
+    Item(t)
+
+  /**
+   * Constructed an optional batch from a collection of values.
+   *
+   * Since batches cannot be empty, this method returns `None` if `ts`
+   * is empty, and `Some(batch)` otherwise.
+   */
+  def items[T](ts: TraversableOnce[T]): Option[Batched[T]] =
+    if (ts.isEmpty) None else {
+      val it = ts.toIterator
+      val t0 = it.next
+      Some(Item(t0).append(it))
+    }
+
+  /**
+   * The free semigroup for batched values.
+   *
+   * This semigroup just accumulates batches and doesn't ever evaluate
+   * them to flatten the tree.
+   */
+  implicit def semigroup[A]: Semigroup[Batched[A]] =
+    new Semigroup[Batched[A]] {
+      def plus(x: Batched[A], y: Batched[A]): Batched[A] = x combine y
+    }
+
+  /**
+   * Compacting semigroup for batched values.
+   *
+   * This semigroup ensures that the batch's tree structure has fewer
+   * than `batchSize` values in it. When more values are added, the
+   * tree is compacted using `s`.
+   */
+  def semigroup[A](batchSize: Int)(implicit s: Semigroup[A]): Semigroup[Batched[A]] =
+    new BatchedSemigroup[A](batchSize, s)
+
+  /**
+   * Monoid for batched values.
+   *
+   * This monoid just accumulates batches and doesn't ever evaluate
+   * them to flatten the tree.
+   *
+   * It represents the identity using `Batched(m.zero)`, but does not
+   * have a unique representation of zero. For example, `zero combine
+   * zero` has a different tree structure to `zero`, but they both
+   * represent the same value (as seen by `.sum`).
+   */
+  def monoid[A](implicit m: Monoid[A]): Monoid[Batched[A]] =
+    new Monoid[Batched[A]] {
+      val zero: Batched[A] = Batched(m.zero)
```

Comment:
okay, I see what is going on... we have two different `Eq[Batched[T]]` we have `a.sum == b.sum` and `a.toList == b.toList`.

This worries me a bit. I wonder if we should use two different type name here (`Concat[T]` and `Batched[T]` for the `.toList` and `.sum` cases?) It is easy to get confused which case we are in.